File: templates/python/algorithms/013-union-find.py
"""
013 - Union-find / disjoint set patterns

Use when the problem asks about dynamic connectivity, grouping components,
cycle detection in undirected graphs, or merging accounts/equivalence classes.
"""


class DSU:
    # Variant 1: path compression + union by size.
    # Example problems: number of connected components, redundant connection.
    # Amortized time: almost O(1) per find/union, commonly written as O(alpha(n)).
    # Space: O(n)
    def __init__(self, n: int) -> None:
        self.parent = list(range(n))
        self.size = [1] * n
        self.components = n

    def find(self, x: int) -> int:
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, a: int, b: int) -> bool:
        root_a = self.find(a)
        root_b = self.find(b)
        if root_a == root_b:
            return False

        if self.size[root_a] < self.size[root_b]:
            root_a, root_b = root_b, root_a

        self.parent[root_b] = root_a
        self.size[root_a] += self.size[root_b]
        self.components -= 1
        return True


# Variant 2: count connected components from edges.
# Example problems: number of connected components in an undirected graph.
# Time: O((n + e) * alpha(n))
# Space: O(n)
def count_components(n: int, edges: list[tuple[int, int]]) -> int:
    dsu = DSU(n)
    for a, b in edges:
        dsu.union(a, b)
    return dsu.components


# Variant 3: detect cycle / redundant edge.
# Example problems: redundant connection, validate tree.
# Time: O(e * alpha(n))
# Space: O(n)
def find_redundant_edge(edges: list[tuple[int, int]]) -> tuple[int, int] | None:
    dsu = DSU(max(max(a, b) for a, b in edges) + 1)

    for a, b in edges:
        if not dsu.union(a, b):
            return a, b

    return None


# Variant 4: union by arbitrary keys using id compression.
# Example problems: accounts merge, sentence similarity, equation satisfiability.
# Time: O((n + e) * alpha(n))
# Space: O(n)
def group_equivalent_items(pairs: list[tuple[str, str]]) -> dict[int, list[str]]:
    ids: dict[str, int] = {}

    def get_id(item: str) -> int:
        if item not in ids:
            ids[item] = len(ids)
        return ids[item]

    for a, b in pairs:
        get_id(a)
        get_id(b)

    dsu = DSU(len(ids))
    for a, b in pairs:
        dsu.union(ids[a], ids[b])

    groups: dict[int, list[str]] = {}
    for item, item_id in ids.items():
        root = dsu.find(item_id)
        groups.setdefault(root, []).append(item)

    return groups

File: templates/golang/algorithms/013-union-find.go
package algorithms

/*
013 - Union-find / disjoint set patterns

Use when the problem asks about dynamic connectivity, grouping components,
cycle detection in undirected graphs, or merging accounts/equivalence classes.
*/

// Variant 1: path compression + union by size.
// Example problems: number of connected components, redundant connection.
// Amortized time: almost O(1) per find/union, commonly written as O(alpha(n)).
// Space: O(n)
type DSU struct {
	parent     []int
	size       []int
	components int
}

func NewDSU(n int) *DSU {
	parent := make([]int, n)
	size := make([]int, n)
	for i := 0; i < n; i++ {
		parent[i] = i
		size[i] = 1
	}
	return &DSU{parent: parent, size: size, components: n}
}

func (d *DSU) Find(x int) int {
	if d.parent[x] != x {
		d.parent[x] = d.Find(d.parent[x])
	}
	return d.parent[x]
}

func (d *DSU) Union(a, b int) bool {
	rootA, rootB := d.Find(a), d.Find(b)
	if rootA == rootB {
		return false
	}
	if d.size[rootA] < d.size[rootB] {
		rootA, rootB = rootB, rootA
	}
	d.parent[rootB] = rootA
	d.size[rootA] += d.size[rootB]
	d.components--
	return true
}

func (d *DSU) Components() int {
	return d.components
}

// Variant 2: count connected components from edges.
// Example problems: number of connected components in an undirected graph.
// Time: O((n + e) * alpha(n))
// Space: O(n)
func CountComponents(n int, edges [][2]int) int {
	dsu := NewDSU(n)
	for _, edge := range edges {
		dsu.Union(edge[0], edge[1])
	}
	return dsu.Components()
}

// Variant 3: detect cycle / redundant edge.
// Example problems: redundant connection, validate tree.
// Time: O(e * alpha(n))
// Space: O(n)
func FindRedundantEdge(edges [][2]int) ([2]int, bool) {
	largest := -1
	for _, edge := range edges {
		largest = max(largest, max(edge[0], edge[1]))
	}
	dsu := NewDSU(largest + 1)
	for _, edge := range edges {
		if !dsu.Union(edge[0], edge[1]) {
			return edge, true
		}
	}
	return [2]int{}, false
}

// Variant 4: union by arbitrary keys using ID compression.
// Example problems: accounts merge, sentence similarity, equation satisfiability.
// Time: O((n + e) * alpha(n))
// Space: O(n)
func GroupEquivalentItems(pairs [][2]string) map[int][]string {
	ids := make(map[string]int)
	items := make([]string, 0)
	getID := func(item string) int {
		if id, ok := ids[item]; ok {
			return id
		}
		id := len(ids)
		ids[item] = id
		items = append(items, item)
		return id
	}

	for _, pair := range pairs {
		getID(pair[0])
		getID(pair[1])
	}
	dsu := NewDSU(len(ids))
	for _, pair := range pairs {
		dsu.Union(ids[pair[0]], ids[pair[1]])
	}

	groups := make(map[int][]string)
	for _, item := range items {
		root := dsu.Find(ids[item])
		groups[root] = append(groups[root], item)
	}
	return groups
}
